rmw_implementation/rmw_implementation-extras.cmake.in

```diff
@@ -15,69 +15,55 @@
-        "'${requested_rmw_implementation}' through the environment variable "
-        "'RMW_IMPLEMENTATION', "
-        "however it is not in the list of supported rmw implementations, "
-        "which was specified when the 'rmw_implementation' package was built.")
-    endif()
-  else()
-    if(NOT requested_rmw_implementation STREQUAL default_rmw_implementation)
-      message(FATAL_ERROR
-        "The RMW implementation has been specified as "
-        "'${requested_rmw_implementation}' "
-        "through the environment variable 'RMW_IMPLEMENTATION', "
-        "however this needs to match the RMW implementation "
-        "'${default_rmw_implementation}', "
-        "which was specified when the 'rmw_implementation' package was built.")
-    endif()
+if(@RMW_IMPLEMENTATION_DISABLE_RUNTIME_SELECTION@)
+  message(STATUS "Using RMW implementation '@RMW_IMPLEMENTATION@'")
+  if(NOT requested_rmw_implementation STREQUAL "@RMW_IMPLEMENTATION@")
+    message(FATAL_ERROR
+      "The RMW implementation has been specified as "
+      "'${requested_rmw_implementation}' via "
+      "${requested_rmw_implementation_from}, but rmw_implementation was built "
+      "only with support for '@RMW_IMPLEMENTATION@'.")
+  endif()
+  find_package("@RMW_IMPLEMENTATION@" REQUIRED)
+
+  # TODO should never need definitions and include dirs?
+  list(APPEND rmw_implementation_DEFINITIONS
+    ${@RMW_IMPLEMENTATION@_DEFINITIONS})
+  list(APPEND rmw_implementation_INCLUDE_DIRS
+    ${@RMW_IMPLEMENTATION@_INCLUDE_DIRS})
+  list(APPEND rmw_implementation_LIBRARIES
+    ${@RMW_IMPLEMENTATION@_LIBRARIES})
+else()
+
+  get_available_rmw_implementations(available_rmw_implementations)
+
+  if(NOT requested_rmw_implementation IN_LIST available_rmw_implementations)
+    message(FATAL_ERROR
+      "The RMW implementation has been specified as "
+      "'${requested_rmw_implementation}' via "
+      "${requested_rmw_implementation_from}, but it is not available at this "
+      "time.\n\nCurrently available middlewares:\n"
+      "'${available_rmw_implementations}'")
   endif()
-endif()
 
-if(@RMW_IMPLEMENTATION_SUPPORTS_POCO@)
   set(selected_rmw_implementation "@PROJECT_NAME@")
   # no need to find_package @PROJECT_NAME@
   # since this code is already part of a find_package call of that package
-else()
-  set(selected_rmw_implementation "${default_rmw_implementation}")
-  message(STATUS "Using RMW implementation '${selected_rmw_implementation}' as default")
-  find_package("${selected_rmw_implementation}" REQUIRED)
 
-  # TODO should never need definitions and include dirs?
-  list(APPEND rmw_implementation_DEFINITIONS ${${selected_rmw_implementation}_DEFINITIONS})
-  list(APPEND rmw_implementation_INCLUDE_DIRS ${${selected_rmw_implementation}_INCLUDE_DIRS})
-  list(APPEND rmw_implementation_LIBRARIES ${${selected_rmw_implementation}_LIBRARIES})
+  find_package(Threads REQUIRED)
+  list(APPEND rmw_implementation_LIBRARIES "${CMAKE_THREAD_LIBS_INIT}")
```

Comment:
Why is this local to the `else` case where it was also used in the `if` case before?

Reply:
You're totally right. I'm not sure that the threading library even belongs in here but removing it breaks the build, so :man_shrugging: